### PR TITLE
 ENH use pyrodigal instead prodigal_sm

### PR DESCRIPTION
**Pull request by celiosantosjr**
Substitute prodigal_sm by pyrodigal to avoid installing locally prodigal and allow the hybrid gene prediction scheme, where it uses pre-trained models in small contigs and train models on large contigs. The code outputs the same results previously outputted by Macrel in the same fashion.

### CODE REVIEW

File: install.sh [hunk 1]
```diff
@@ -29,41 +29,43 @@ BASEDIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
 
 echo "# Creating new environment for Macrel"
 mkdir -p envs
-conda create --yes -p $BASEDIR/envs/Macrel_env python=$PYTHON_VERSION
+conda create --yes -p $BASEDIR/envs/Macrel_env python=$PYTHON_VERSION ## Fixed
```
Comment (luispedro) [inline install.sh:32]: What does `# Fixed` mean?

File: install.sh [hunk 2]
```diff
@@ -29,41 +29,43 @@ BASEDIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
...
 echo "# Creating new environment for Macrel"
 mkdir -p envs
-conda create --yes -p $BASEDIR/envs/Macrel_env python=$PYTHON_VERSION
+conda create --yes -p $BASEDIR/envs/Macrel_env python=$PYTHON_VERSION ## Fixed 
+source activate $BASEDIR/envs/Macrel_env
 conda config --env --add channels defaults
 conda config --env --add channels bioconda
 conda config --env --add channels conda-forge
 
+echo "# Installing packages with MAMBA"
```
Comment (luispedro) [inline install.sh:38]: This is incorrect. It may or may not use mamba

File: install.sh [hunk 3]
```diff
@@ -29,41 +29,43 @@ BASEDIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
...
+              megahit \
+              paladin \
+              pandas \
+              scikit-learn \
+              atomicwrites \
+              rpy2 \
+              tzlocal \
+              --quiet
 fi
-${CONDA_INSTALL_CMD} install -y \
```
Comment (luispedro) [inline install.sh:43]: This seems like a pessimization, completely breaking DRY.

File: install.sh [hunk 4]
```diff
@@ -29,41 +29,43 @@ BASEDIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
...
-# env forwards the current environment and sets some extra variables needed by
-# ./recipe/build.sh
 env \
         PYTHON=$(which python) \
         SRC_DIR=$PWD \
         PREFIX=$CONDA_PREFIX \
-        ./recipe/build.sh
-
-
+        sh ./recipe/build.sh
```
Comment (luispedro) [inline install.sh:67]: The file uses `#!/usr/bin/bash` not `sh`. Not sure if it matters, but should be consistent

File: recipe/build.sh [hunk 5]
```diff
@@ -2,14 +2,5 @@
...
-mkdir -p $PREFIX/bin
-
-cd prodigal_modified
-make CC=$CC # conda will add $CC to environment
-chmod +x prodigal
-cd ..
-
-cp $SRC_DIR/prodigal_modified/prodigal $PREFIX/bin/prodigal_sm
-
+$PYTHON -m pip install pyrodigal>=0.7.3
```
Comment (luispedro) [inline recipe/build.sh:5]: Why is this not installed with `conda`/`mamba`?

File: tests/get-smorfs/expected.smorfs.w_headers.faa [hunk 6]
```diff
@@ -10,19 +10,21 @@ MLFQNDDMPMDKYQIGIPQPYMDDDRKWVLTYGISCRPLMPSAAKRLLSKTPLSIDKALVQELKISKEDIRLMKHYNMFV
 MAMGIVAFGFGTAGGVLLAKLMNLFLKEKINPLIGSAGVSAVPMAARVSQVVGQKENPSNFLLMHAMGPNVAGVIGSAIAAGVLMSLFG*
 >scaffold95393_2_MH0058_2 # 542 # 826 # 1 # ID=5_2;partial=00;start_type=ATG;rbs_motif=AGGA;rbs_spacer=5-10bp;gc_cont=0.372
 MTTQTRAQQLKEIEFQTQMLNNLKKWIRNLIILSSIGIILAYWGLGVQSKMPFTVFGVAGVIITIISVILCVVIGLGIKRGRANVDKILQLVKA*
->scaffold8449_1_MH0058_1 # 119 # 247 # -1 # ID=6_1;partial=00;start_type=ATG;rbs_motif=GGAGG;rbs_spacer=5-10bp;gc_cont=0.550
+>scaffold8449_1_MH0058_1 # 119 # 247 # -1 # ID=6_1;partial=00;start_type=ATG;rbs_motif=GGAGG;rbs_spacer=5-10bp;gc_cont=0.55
```
Comment (luispedro) [inline tests/get-smorfs/expected.smorfs.w_headers.faa:13]: Can we keep the same `gc_cont` format as the previous version?

File: tests/get-smorfs/expected.smorfs.w_headers.faa [hunk 7]
```diff
@@ -10,19 +10,21 @@ MLFQNDDMPMDKYQIGIPQPYMDDDRKWVLTYGISCRPLMPSAAKRLLSKTPLSIDKALVQELKISKEDIRLMKHYNMFV
...
 >scaffold10455_3_MH0058_2 # 292 # 504 # -1 # ID=7_2;partial=00;start_type=ATG;rbs_motif=GGAGG;rbs_spacer=3-4bp;gc_cont=0.648
 MFRRISRQERERRAARAELETTLQALRSNERAFTEAQDPFYIDQLTYQHAALMCRCRALLRTLRAEGEEP*
 >scaffold10455_3_MH0058_3 # 515 # 787 # -1 # ID=7_3;partial=00;start_type=ATG;rbs_motif=None;rbs_spacer=None;gc_cont=0.608
 MRANIHTELSPTVGSEQQGEELIKSLLADFETDPGKLWSTNIFGKSLNELVSEGVQAKLLHMPQEARSRLQQTLERIINEGCDGLICILL*
+>scaffold98598_5_MH0058_1 # 1466 # 1756 # -1 # ID=8_1;partial=00;start_type=ATG;rbs_motif=GGA/GAG/AGG;rbs_spacer=5-10bp;gc_cont=0.378
+MGVCLVVFGVLFLVHIFGGFLNYELIFHLWPLILIGLGMEILFYTCIASGKAEKYDFAAIIVIFLLTGFSMCMAGADWVITNVPQDVWFNSTGYGF*
 >scaffold98598_5_MH0058_4 # 3080 # 3358 # -1 # ID=8_4;partial=00;start_type=ATG;rbs_motif=AGxAGG/AGGxGG;rbs_spacer=5-10bp;gc_cont=0.516
 MPSESEVLESRAGTVSSPEERFLEEEVKDELAKACRDLKPPYGQIAYEFYCCDRTAREIAEQKGIKLKTVQTQVLRARKMLQKKLGPKIVRP*
->scaffold76045_5_MH0058_1 # 44 # 259 # 1 # ID=9_1;partial=00;start_type=ATG;rbs_motif=None;rbs_spacer=None;gc_cont=0.500
+>scaffold76045_5_MH0058_1 # 44 # 259 # 1 # ID=9_1;partial=00;start_type=ATG;rbs_motif=None;rbs_spacer=None;gc_cont=0.5
```
Comment (luispedro) [inline tests/get-smorfs/expected.smorfs.w_headers.faa:25]: as above

File: macrel/ORFs_prediction.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+                predictions.append(ppyrodigal_out(h, i+1, idx+1, pred))
+        else:
+            # if contig length is above or 100kbp then suitable for training of
+            # its own model, therefore proceed in a genome wise way
+            # each train procedure updates the predictor automatically
+            gorf.train(s)
+            for i, pred in enumerate(gorf.find_genes(s)):
+                predictions.append(ppyrodigal_out(h, i+1, idx+1, pred))
+    # saving sequences
+    with open(ofile, 'w') as odb:
```
Comment (luispedro) [inline macrel/ORFs_prediction.py:51]: Why not use `atomicwrites`?

File: macrel/ORFs_prediction.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    header = f'>{orfid} # {pred.begin} # {pred.end} # {pred.strand} # {last}'
+    return (header, pred.translate())
+
+
+def predict_genes(infile, ofile):
+    import random
+    import pandas as pd
+    from .fasta import fasta_iter
+    random.seed(1991)
+    predictions = []
```
Comment (luispedro) [inline macrel/ORFs_prediction.py:34]: Generally speaking, I'm not a big fan of this type of structure: why not directly write to the output (particularly, if using `atomicwrites`)?

File: setup.py [hunk 10]
```diff
@@ -73,7 +72,7 @@
       install_requires=[
           'scikit-learn',
           'pandas',
-          'atomicwrites',
+          'atomicwrites'
```
Comment (luispedro) [inline setup.py:75]: why is this necessary?

**Review by luispedro — CHANGES_REQUESTED** [review]

File: install.sh (hunk 2 above)
Comment (celiosantosjr) [reply install.sh]: It was just misplaced, should be bellow

File: install.sh (hunk 3 above)
Comment (celiosantosjr) [reply install.sh:43]: corrected it

File: install.sh (hunk 4 above)
Comment (celiosantosjr) [reply install.sh]: it was corrected now

File: macrel/ORFs_prediction.py (hunk 9 above)
Comment (celiosantosjr) [reply macrel/ORFs_prediction.py]: implemented as asked

File: macrel/ORFs_prediction.py (hunk 8 above)
Comment (celiosantosjr) [reply macrel/ORFs_prediction.py]: done

File: recipe/build.sh (hunk 5 above)
Comment (celiosantosjr) [reply recipe/build.sh]: moved to the install script

**Review by celiosantosjr — COMMENTED** [review]
Fixed the previously requested changes in these new three commits

**Review by luispedro — CHANGES_REQUESTED** [review]
I think it's almost there. I still think the changes to the `mamba`/`conda` call are clearly worse than before and an unnecessarily large change (why not a 1-line change to add pyrogal to the extant list of dependencies?)

Documentation is completely missing, though. At the very least, a short update to the `ChangeLog` and a longer one for https://macrel.readthedocs.io/en/latest/whatsnew/ extolling the improvements.

File: install.sh [hunk 11]
```diff
@@ -30,40 +30,46 @@ BASEDIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
...
 conda create --yes -p $BASEDIR/envs/Macrel_env python=$PYTHON_VERSION
+
+source activate $BASEDIR/envs/Macrel_env
 conda config --env --add channels defaults
 conda config --env --add channels bioconda
 conda config --env --add channels conda-forge
 
 if ! which mamba > /dev/null; then
     CONDA_INSTALL_CMD=conda
+    conda install -y \
```
[the remark this answers is not in the document]
Comment (celiosantosjr) [reply install.sh]: Reverted to the previously used install.sh, I used a local version of the macrel which had it wrong.
